src/ucp/wireup/wireup_cm.c

```diff
@@ -194,7 +194,85 @@ static void ucp_cm_client_connect_cb(uct_ep_h ep, void *arg,
                                      const uct_cm_remote_data_t *remote_data,
                                      ucs_status_t status)
 {
-    ucs_error("UCP CM wireup is not completely implemented");
+    ucp_ep_h                   ucp_ep     = (ucp_ep_h)arg;
+    ucp_worker_h               worker     = ucp_ep->worker;
+    ucp_wireup_ep_t            *wireup_ep = ucp_ep_get_cm_wireup_ep(ucp_ep);
+    unsigned                   addr_indices[UCP_MAX_RESOURCES];
+    ucp_ep_config_key_t        key;
+    ucp_unpacked_address_t     addr;
+    ucp_wireup_sockaddr_data_t *sa_data;
+    unsigned                   addr_idx;
+    ucp_lane_index_t           lane;
+
+    if (status != UCS_OK) {
+        goto out;
+    }
+
+    ucs_assert(wireup_ep != NULL);
+    sa_data = (ucp_wireup_sockaddr_data_t *)remote_data->conn_priv_data;
+    ucp_ep_update_dest_ep_ptr(ucp_ep, sa_data->ep_ptr);
+    status = ucp_address_unpack(worker, sa_data + 1,
+                                UCP_ADDRESS_PACK_FLAG_IFACE_ADDR |
+                                UCP_ADDRESS_PACK_FLAG_EP_ADDR, &addr);
+    if (status != UCS_OK) {
+        goto out;
+    }
+
+    if (addr.address_count == 0) {
+        status = UCS_ERR_UNREACHABLE;
+        goto free_addr_list;
+    }
+
+    ucs_assert(addr.address_count <= UCP_MAX_RESOURCES);
+    for (addr_idx = 0; addr_idx < addr.address_count; ++addr_idx) {
+        addr.address_list[addr_idx].dev_addr = remote_data->dev_addr;
+    }
+
+    /* re-configure lanes according to real remote address  */
+    ucp_ep_config_key_reset(&key);
+    ucs_assert(wireup_ep->ep_init_flags & UCP_EP_INIT_CM_WIREUP_CLIENT);
+    status = ucp_wireup_select_lanes(ucp_ep, wireup_ep->ep_init_flags,
+                                     ucp_ep_get_tl_bitmap(ucp_ep), &addr,
+                                     addr_indices, &key);
+    if (status != UCS_OK) {
+        goto free_addr_list;
+    }
+
+    status = ucp_worker_get_ep_config(worker, &key, 0, &ucp_ep->cfg_index);
+    if (status != UCS_OK) {
+        goto free_addr_list;
+    }
+
+    /* establish connections on all underlying endpoints */
+    for (lane = 0; lane < ucp_ep_num_lanes(ucp_ep); ++lane) {
+        if (lane == ucp_ep_get_cm_lane(ucp_ep)) {
+            continue;
+        }
+
+        status = ucp_wireup_connect_lane(ucp_ep, wireup_ep->ep_init_flags, lane,
+                                         &addr, addr_indices[lane]);
+        if (status != UCS_OK) {
+            goto free_addr_list;
+        }
+    }
+
+    status = ucp_wireup_resolve_proxy_lanes(ucp_ep);
+    if (status != UCS_OK) {
+        goto free_addr_list;
+    }
+
+    ucp_wireup_connect_local(ucp_ep, &addr, NULL);
+
+    ucp_wireup_remote_connected(ucp_ep);
+    ucp_ep_flush_state_reset(ucp_ep);
```

Comment:
it could be done in ucp_ep_create_to_sock_addr(), because for client/server we don't go through match_ctx, so there is no union conflict with flush_state